### PR TITLE
feat(webpack-dev-server): add angular handler

### LINKED ISSUE #22049: Add Support for Angular Component Testing
### What would you like?

With the release of `10.0` we now have Beta support for both `Vue` and `React`. We need to also add support for `Angular`.

### Why is this needed?

This is needed to support our wonderful Angular users who want to do Cypress Component Testing in their Angular applications.

### Other

_No response_

### LINKED ISSUE #22812: Add Angular handler to @cypress/webpack-dev-server
In order to support Angular for Component Testing, we need to add an Angular handler to our `npm/webpack-dev-server` package. We are targeting v13/v14 support.

The responsibilities of this handler are:
- Source the Webpack related dependencies needed to compile (webpack, webpack-dev-server, html-webpack-plugin)
- Source a Webpack config that will compile Angular components

This handler will be targeted at users that created Angular projects using the `@angular/cli` package. If a user is not using this package, they will most likely have configured their own `webpack.config.js` that they can pass in. This handler will be similar to the other handlers we have for higher-order frameworks `{create-react-app, next, nuxt, vue-cli}`.

Angular offers a Webpack plugin for compiling Angular projects: https://www.npmjs.com/package/@ngtools/webpack
The `@angular/devkit` has internal APIs for sourcing an `@angular/cli` specific project config, though it is contextual to their `builder` API so it isn't as straightforward to use these APIs: https://github.com/angular/angular-cli/blob/main/packages/angular_devkit/build_angular/src/builders/browser/index.ts#L109

Ideally, it would be great to use the `@angular/devkit` APIs as doing so would allow us to respect the options specified by the user (normally stored in a project's `angular.json`). Otherwise, a custom solution that fits most of the needs of the user is acceptable.

### Requirements:

- Source webpack deps and webpack config that enables the compilation of Angular components
- System tests in `npm/webpack-dev-server` proving that the compilation is successful (There is an [external Angular mounting package](cypress-angular-component-testing) we can utilize while we work on official mounting support)

There is an in-progress PR for this work: #22314

### PR DESCRIPTION
**Pull request by jordanpowell88**
- Closes #22049
- Closes #22812

### User facing changelog
Adds Angular framework support for DevServer.

### Additional details
This dev-server is based on `@angular/cli` created angular projects. It sources the webpack config from the same build process used to build angular/cli projects which allows us to respect the build options that users have in their `angular.json` and tsconfig.

### Steps to test
There are system tests in `npm/webpack-dev-server` that run against an Angular 13 and 14 application. To test this on an Angular application outside of system tests, reference the `system-tests/project-fixtures/angular` angular project fixture. Since we don't have `mount` merged, there is a local version that should be used instead.

### How has the user experience changed?

https://user-images.githubusercontent.com/25158820/180316864-0a8f938b-d172-4f5a-ad61-c9a3aa98aaa3.mov

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [x] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Comment by lmiller1990** [comment]
@jordanpowell88 should I review this now, or do we want to get CI passing first?

**Comment by jordanpowell88** [comment]
> @jordanpowell88 should I review this now, or do we want to get CI passing first?
> 
> Excited to see Angular support is about to land!

Sorry I was OOO. Originally I was intending on doing 2 separate PR's. 1 for dev server and another for the mount but I feel like it can be 1 PR now so hold off until I get the `mount` code added. I will le tyou know when it is ready

File: npm/webpack-dev-server/src/helpers/angularHandler.ts [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    defaultProject = Object.keys(angularJson.projects).find((name) => angularJson.projects[name].projectType === 'application')
+
+    if (!defaultProject) {
+      throw new Error('Could not find a project with projectType "application" in "angular.json"')
+    }
+  }
+
+  const defaultProjectConfig = angularJson.projects[defaultProject]
+
+  const tsConfig = await generateTsConfig(devServerConfig, defaultProjectConfig)
```
Comment (ZachJW34) [inline npm/webpack-dev-server/src/helpers/angularHandler.ts:62]: We are generating a tmp tsConfig but another option would be to scaffold one to disk for them. There might be use cases for allowing the user to extend types or add TS files to the compilation which having one that the user owns would enable. Scaffolding it makes it part of our API, so if we want to make changes to it in the future it might be a breaking change. We could also provide an option for passing in a custom tsConfig.

File: npm/webpack-dev-server/src/helpers/angularHandler.ts [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    (wco: any) => [getCommonConfig(wco), getStylesConfig(wco)],
+  )
+
+  delete config.entry.main
+
+  return config
+}
+
+export function getAngularBuildOptions (projectConfig: AngularJsonProjectConfig, tsConfig: string) {
+  return {
```
Comment (ZachJW34) [inline npm/webpack-dev-server/src/helpers/angularHandler.ts:80]: These options are derived from https://github.com/angular/angular-cli/blob/main/packages/angular_devkit/build_angular/src/builders/browser/schema.json

A comment would be useful here.

File: npm/webpack-dev-server/src/helpers/angularHandler.ts [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+export async function getTempDir (): Promise<string> {
+  const cypressTempDir = path.join(tmpdir(), 'cypress-angular-ct')
+
+  await fs.ensureDir(cypressTempDir)
+
+  return cypressTempDir
+}
+
+export async function getAngularCliModules (projectRoot: string) {
```
Comment (ZachJW34) [inline npm/webpack-dev-server/src/helpers/angularHandler.ts:172]: Angular 13 and 14 share the same API, but if we wanted to support Angular 12 we would need to change this. Need to look into current Angular 12 usages.

File: npm/webpack-dev-server/src/makeDefaultWebpackConfig.ts [hunk 4]
```diff
@@ -37,9 +37,9 @@ export function makeDefaultWebpackConfig (
     },
     plugins: [
       new HtmlWebpackPlugin({
-        // Todo: Add indexHtmlFile when it gets added as a config property
         template: indexHtmlFile,
-      }) as any,
+        ...(config.devServerConfig.framework === 'angular' ? { scriptLoading: 'module' } : {}),
```
Comment (ZachJW34) [inline npm/webpack-dev-server/src/makeDefaultWebpackConfig.ts:41]: Angular generates all of it's scripts with `<script type="module">`. Without this option, live-reloading breaks.

File: npm/webpack-dev-server/src/makeWebpackConfig.ts [hunk 5]
```diff
@@ -72,6 +72,29 @@ function modifyWebpackConfigForCypress (webpackConfig: Partial<Configuration>) {
   return webpackConfig
 }
 
+async function addEntryPoint (webpackConfig: Configuration) {
```
Comment (ZachJW34) [inline npm/webpack-dev-server/src/makeWebpackConfig.ts:75]: Angular includes global styles and polyfills via the `webpackConfig.entry` field. We were overriding the entry with our Cypress entrypoint, so I refactored the handling of entrypoints to preserve additional entrypoints and deleted unneeded entrypoints from the other framework handlers.

File: system-tests/project-fixtures/angular/src/app/svg.component.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import { Component } from "@angular/core";
```
Comment (ZachJW34) [inline system-tests/project-fixtures/angular/src/app/svg.component.ts:1]: Can delete `svg.component.ts`, this what just a test to verify that svg templates load. It's a react svg anyway 😆

File: system-tests/project-fixtures/angular/cypress.config.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+export default defineConfig({
+  component: {
+    devServer: {
+      framework: 'angular',
+      bundler: 'webpack',
+      webpackConfig: {
+        resolve: {
+          alias: {
+            '@angular/common': require.resolve('@angular/common'),
```
Comment (ZachJW34) [inline system-tests/project-fixtures/angular/cypress.config.ts:11]: Same issue we've had with other framework system tests. These aliases are not required when using an external application

File: npm/webpack-dev-server/src/devServer.ts [hunk 8]
```diff
@@ -22,10 +23,10 @@ export type WebpackDevServerConfig = {
   onConfigNotFound?: (devServer: 'webpack', cwd: string, lookedIn: string[]) => void
 } & {
   framework?: typeof ALL_FRAMEWORKS[number] // Add frameworks here as we implement
-  webpackConfig?: unknown // Derived from the user's webpack
+  webpackConfig?: unknown // Derived from the user's webpack,
```
Comment (lmiller1990) [inline npm/webpack-dev-server/src/devServer.ts:26]: Typo? `,`

File: npm/webpack-dev-server/src/helpers/createReactAppHandler.ts [hunk 9]
```diff
@@ -77,6 +77,8 @@ function loadWebpackConfig (devServerConfig: WebpackDevServerConfig): Configurat
       webpackConfig = webpackConfig('development')
     }
 
+    delete webpackConfig.entry
```
Comment (lmiller1990) [inline npm/webpack-dev-server/src/helpers/createReactAppHandler.ts:80]: How come we need this? Seems unrelated to Angular. Drive by bug fix?

Comment (lmiller1990) [reply npm/webpack-dev-server/src/helpers/createReactAppHandler.ts:80]: Right: https://github.com/cypress-io/cypress/pull/22314/files#r927118026

This must have been a bit painful to figure out.

**Review by lmiller1990 — APPROVED** [review]
Couple minor comments/questions - tentative approval. I could pull this down but it might make more sense to merge, add launchpad integration, then give that a test like a production-like user?

File: npm/webpack-dev-server/src/helpers/angularHandler.ts (hunk 3 above)
Comment (lmiller1990) [reply npm/webpack-dev-server/src/helpers/angularHandler.ts:172]: 12.2 (most popular 12.x version)

image

Seems on par with 13.x and 14.x Definitely good to ship with some and look to add more if needed - we can set in launchpad what version(s) we support, for now >= ^13 <= ^14

File: npm/webpack-dev-server/src/helpers/createReactAppHandler.ts (hunk 9 above)
Comment (ZachJW34) [reply npm/webpack-dev-server/src/helpers/createReactAppHandler.ts:80]: Yes I had to refactor the entrypoints of other handlers to get this to work. I could have had an Angular specific hack here, but I wanted to minimize framework specific logic here as this is common to all.

File: npm/webpack-dev-server/src/devServer.ts (hunk 8 above)
Comment (ZachJW34) [reply npm/webpack-dev-server/src/devServer.ts:26]: I added an option here and then removed it, good catch

File: cli/react18/package.json [hunk 10]
```diff
@@ -0,0 +1,57 @@
+{
```
Comment (astone123) [inline cli/react18/package.json:1]: Why is this in the diff?

**Review by astone123 — APPROVED** [review]
Looks like there might be something up with a part of the diff. Other than that, looks good 👍🏻

File: cli/react18/package.json (hunk 10 above)
Comment (ZachJW34) [reply cli/react18/package.json:1]: No clue, should be .gitignored. Will update!

Comment (ZachJW34) [reply cli/react18/package.json]: Addressed in 82b609cf5abab8e111ad1f4ce592bb0be79cc4cd
